### PR TITLE
No MMI Test For You

### PR DESCRIPTION
**Pull request by DatBoiTim**
## About The Pull Request
Kills people who MMI Test Lings.

Its about sending a message. Ling players who abuse this should similarly be banned for banbaiting because this does notify admins and will likely result in a bwoink to the person with the MMI.

## Why It's Good For The Game
Code Oversight for metagaming bad

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

GibbedLMAO
Rulebreakers will be gibbed. Laugh at them. When you're done laughing, ban them.

</details>

## Changelog
:cl: DatBoiTim
tweak: Added a Check to MMIs for ling testing. If you ling test someone you get brutally murdered in an immersive way before admins get notified.
/:cl:

### CODE REVIEW

**Comment by Tsar-Salat** [comment]
Honestly, I was leaning towards dusting to round remove the cringe. 

This works too though

**Comment by DatBoiTim** [comment]
> Honestly, I was leaning towards dusting to round remove the cringe.
> 
> This works too though

Doesn't send enough of a message IMO.

Dusting is great for permanently round removing someone yes, but seeing someone reduced to a pile of gore after a rulebreak really sends a message.

File: code/modules/mob/living/brain/MMI.dm [hunk 1]
```diff
@@ -42,6 +42,12 @@
 		if(brain)
 			to_chat(user, "<span class='warning'>There's already a brain in the MMI!</span>")
 			return
+		if(newbrain.decoy_override) //No MMI Test For You
+			to_chat(user, "<span class='warning'>There's a decent chance this is against the rules, admins have been notified and you're getting gibbed.</span>")
+			user.gib()
+			message_admins("[key_name(user)] was gibbed for trying to ling test someone with an MMI")
```
Comment (mystery3525) [inline code/modules/mob/living/brain/MMI.dm]:
```suggestion
			message_admins("[key_name(user)] tried to ling test someone with an MMI")
```

Comment (mystery3525) [reply code/modules/mob/living/brain/MMI.dm]: I'm against using code to enforce policy, especially in this way, its a serious immersion-break when someone just explodes into a million if they put what they think is a brain into a machine; but you CAN totally add logging for stuff like this

Comment (Rukofamicom) [reply code/modules/mob/living/brain/MMI.dm]: I don't disagree that it shouldn't happen in such an immersion breaking way way, but code should always enforce policy when possible. 

The no-ling testing policy is absolute - there is never a time when it is allowed, so it is a perfect candidate for something that should just be outright prevented by code.

File: code/modules/mob/living/brain/MMI.dm [hunk 2]
```diff
@@ -42,6 +42,12 @@
...
+			return
+
```
Comment (Rukofamicom) [inline code/modules/mob/living/brain/MMI.dm]:
```suggestion
		if(newbrain.decoy_override)
			user.visible_message("<span class='warning'>[newbrain] contorts violently, whipping a sharp skewer through [user]'s head!</span>", "<span class='warning'>[newbrain] contorts violently, whipping a sharp skewer through your head!</span>", "<span class='italics'>You hear organic matter ripping and tearing!</span>")
			to_chat(user, "<span class='userdanger'>Performing changeling tests with an MMI is not allowed on this server. Admins have been notified</span>")
			message_admins("[key_name(user)] was removed from the round for trying to place a changeling brain into an MMI")
			playsound(newbrain, 'sound/magic/demon_consume.ogg', 50, 1)
			if(iscarbon(user))
				var/mob/living/carbon/C = user
				var/obj/item/bodypart/affecting
				affecting = C.get_bodypart(BODY_ZONE_HEAD)
				C.adjustOrganLoss(ORGAN_SLOT_BRAIN, 200)
				affecting.receive_damage(150)
			else //Failsafe, but should never almost never trigger
				user.gib()
			user.ghostize(FALSE) //Player is removed from the round permanently, even if they receive medical care.
			return

```

image

Here's a much more immersive way to accomplish the same goal.

**Review by Rukofamicom — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/brain/MMI.dm (hunk 1 above)
Comment (DatBoiTim) [reply code/modules/mob/living/brain/MMI.dm]: I'm with Ruko on this as far as round removing people who do stuff like this. MMI testing is an exploit resulting from a code oversight. 

That's why it became a "admin issue". While there is no 100% fix without completely reworking lings, an exploit is something that can be addressed with both code and admin work.

I'll take his option, if it means we make progress on getting this merged.

**Comment by tontyGH** [comment]
"Oh damn this dude's body is super fucked up lemme put his brain in an MMI so I can ask him what--"

**Comment by mystery3525** [comment]
Or changelings maliciously asking to be borged :trollface:

**Comment by Tsar-Salat** [comment]
> Oh damn this dude's body is super fucked up lemme put his brain in an MMI so I can ask him what--"

A down and out ling has either:
1. Just been killed by sec and is getting dragged to incinerator
2. Doesn't know their basic ling powers
3. Is trying to abuse this feature 

There are not many instances where a ling would bother staying dead, and most doctors aren't gonna mmi you. They are gonna try to patch you up for 10 minutes in surgery then clone you.

**Comment by Rukofamicom** [comment]
> to see if it is an actual issue worth addressing

It 100% is though @mystery3525, let me go over some points:
* Bee used to have actual ling tests in its code that got removed
* Other servers (TG for instance) freely allow players to utilized mmi testing
* We have more than one thread about ling testing on our own forums, one literally made this year trying to argue (in good faith!) that admins are dumb for acting like MMI testing is against the rules. 

It might be *very* clear to you as an administrator of this server that MMI testing is not permissible, but it is only apparent to players after they have been told. An automated system of telling them and actively discouraging it is the best way to accomplish this.

**Comment by tontyGH** [comment]
Okay real input to this, tl;dr this is the worst band-aid fix I have ever seen in my life

Permanently killing someone for trying to put a fake brain in an MMI is not only an incredibly unimmersive and excessive punishment, all this does is punish people who don't know either don't know about the no testing rules (like ruko mentionned) or **don't know that the brain is from a ling at all**, which you hope they wouldn't as the changeling

All this does is tip off the surrounding (now annoyed from having their coworker RRed) staff that the brain is from a changeling, their cover is blown, and we're back to square one, this does nothing but fuck up one persons round

> Most doctors won't MMI a ling

Just because most don't do it doesn't mean that the people that do it should just suck up having their round being **permanently ended** just for trying it. It's not even the only context for someone to be MMI'd, there's also forced cyborgization for example

Are the ruined rounds of well-intentioned people worth it to catch the one guy who actually wanted to metagame?

And what is this punishment? This is lazy and uncreative, why not make the MMI itself explode? Or make it emit a screech to stun everyone and let the ling escape? Or **literally anything from this**

Go in and fix the MMI testing method itself if you want to stop MMI testing this is Not it

**Comment by tontyGH** [comment]
But this is really funny so merge this

**Comment by Penwin0** [comment]
I don't like this PR, I feel like the cure is worse than the poison here, it's really inmersion breaking, I can easily see it being abused or being triggered on accident, doesn't even prevent ling tests at all, you just need one more person as a sacrifice now.

**Comment by LodedDiper69** [comment]
Doing ling tests to exploit the game code are more immersion breaking than this PR ever would be. I'm giving this the unofficial open key 🔑

**Comment by tontyGH** [comment]
damn I got pissed wtf

**Comment by Fronsis** [comment]
Yeah i think people missed the point this is a half-joke half-wake up call pr to give in  a big bold caps a message that says : " MMI TEST IS AGAINST THE RULES STOP" Though at this point i'd said the majority of our players on the sec/command area is aware

It's true that sometimes it feels like you need a method to get the "haha gotcha ling!!!"  But if the player(ling) played their cards correctly they shouldn't be outed by a single "easy" process

So i always have mixed feelings wether we need a system to help identify a changeling or not. It depends a lot on each shift tbh

**Comment by Rukofamicom** [comment]
> maybe a message on #announcement would be enough to call it a day or add it to the relevant wiki.

The majority of not only our players, but all game players do not interact with the the community on the level dedicated players do

This would raise awareness but not on the same level as this PR. I don't think it's immersion breaking in the sense that the changeling can shapeshift and control its body in magical ways.

The forced round removal would affect people, at most, once in their playtime and serve to function as a powerful lesson that what they attempted is not allowed on this server.

**Comment by Rukofamicom** [comment]
I can see the reasoning that this is too extreme, but it absolutely does need to be at least q huge red message for players attempting it at a minimum, even if it doesn't harm or round remove them.

It needs to be very clear to the person attempting that it is not allowed to do what they are attempting in that moment **WITHOUT** needing an admin present. This is not an issue that can wait for an admin to intervene.

If it breaks their immersion for legitimately attempting to do a long test - **BOO FUCKING HOO, THEY WERE ALREADY METAGAMING**.

**Comment by PowerfulBacon** [comment]
I think this is far too verbose

**Comment by itsmeow** [comment]
yeah that's the right call here

**Pull request closed without merging**